### PR TITLE
Internal buffer trading quotes

### PR DESCRIPTION
**Pull request by MartinquaXD**
# Description
See linked #2345 for more context about this change.

# Changes
Added the needed machinery to let external solvers internalize interactions for quotes.

Moved trusted tokens arguments into the shared arguments so `autopilot` and `orderbook` both have access to them.
Updated code that builds the `AutoUpdatingTokenList` and passes it around.

Slightly adjusted the `local_node_buffers` e2e test to see that we can compute quotes based on internal buffer trades.

## How to test
Updated existing e2e test to make sure the new format doesn't cause issues.
In a future PR we could try to add internalization in the baseline logic and verify that the existing e2e test for internalization behaves accordingly.

### CODE REVIEW

File: crates/shared/src/price_estimation/trade_verifier.rs
Comment (fleupold) [inline crates/shared/src/price_estimation/trade_verifier.rs:242]: I don't think the autopilot should decide whether or not to simulate with internalized interactions or not on a case by case basis (but instead leave this logic up to the solvers and simply tell them the settlement contract balance and trusted tokens like for regular solutions). 

The only question IMO should be if we simulate the internalized or uninternalized calldata (I agree with you it's good to do the former to show more competitive prices).

File: crates/shared/src/price_estimation/trade_verifier.rs
Comment (fleupold) [inline crates/shared/src/price_estimation/trade_verifier.rs:103]: The solver may not want all interactions to be internalised (e.g. PMMs) and thus we might internalize things here which later won't be internalizable in the solution we receive during solving.

**Review by fleupold — CHANGES_REQUESTED** [review]
Maybe I'm understanding things wrongly, but I think we should simply treat the solvers response like the response they will later send on `/solve` and simulate it exactly the same way we will later do in the driver/autopilot.

File: crates/shared/src/price_estimation/trade_verifier.rs
Comment (MartinquaXD) [reply crates/shared/src/price_estimation/trade_verifier.rs:242]: Seems like a reasonable solution which will probably result in the `/solve` logic to align closer with the `/quote` logic.

> Maybe I'm understanding things wrongly, but I think we should simply treat the solvers response like the response they will later send on /solve and simulate it exactly the same way we will later do in the driver/autopilot.

That would be the easiest solution but this would also require things like approvals and `ETH` unwraps to happen before a user can even get a verified quote.
These are not a concerns when simulating actual settlements because there we only want to know whether a settlement works right now as is and not theoretically under some additional conditions.

Comment (fleupold) [reply crates/shared/src/price_estimation/trade_verifier.rs:242]: > Also the user didn't sign any order that could be executed during a regular `eth_call`.

Maybe I'm not fully understanding how verified quotes work on a technical level, but in order to simulate a user having signed their order, I think we should be able to use `StorageAccessible.simulateDelegatecall` [[1](https://github.com/cowprotocol/contracts/blob/main/src/contracts/mixins/StorageAccessible.sol)] to simulate having set the presignature of a given order ID (making sure that the raw settlement transaction should pass)

Comment (MartinquaXD) [reply crates/shared/src/price_estimation/trade_verifier.rs:242]: I explained the idea in detail in the original [PR](https://github.com/cowprotocol/services/pull/1575). The TL;DR is that we use state overrides to put a helper contract at the trader's `from` address. With the trader being turned into a contract we control we mainly use it for 3 things:
* make it do stuff the user would normally have to do to prepare the trade (e.g. wrapping and approvals)
* fake the signature by always returning true from `isValidSignature()`
* proxy any unexpected function call to the original code at the `from` address (in case it was actually a smart contract)

So while your suggestion might be used for faking a signature I think we'd still have the problem of the missing approvals and unwraps. And if we have to impersonate the trader anyway I think we might as well keep the `isValidSignature()` trickery because it's trivial by then.

That being said I have very little experience with smart contracts so it could very well be that there is a better solution for this.

File: crates/driver/src/boundary/quote.rs
Comment (fleupold) [inline crates/driver/src/boundary/quote.rs:45]: This could be a separate PR (changing the driver `/quote` logic)

File: crates/shared/src/price_estimation/trade_verifier.rs
Comment (fleupold) [inline crates/shared/src/price_estimation/trade_verifier.rs:95]: Why is this logic in `shared`? Can’t it be - like for solve requests - the driver that decides which interactions to internalise? In fact I’m not even sure why we are using a different route for quotes? To me a quote is a single order solve request (with insta reveal).

**Review by fleupold — COMMENTED** [review]
This PR is getting quite huge and hard to review. I'm also still not convinced about the approach. Why doesn't the driver `/quote` endpoint send us an encoded settlement that we can interpret in the API (or at least filter out interactions it doesn't believe it needs). This way the logic is as close as possible to what we do for `solve` requests. I find it very weird that we filter internalizable interactions in the API.

File: crates/shared/src/price_estimation/trade_verifier.rs
Comment (MartinquaXD) [reply crates/shared/src/price_estimation/trade_verifier.rs:95]: The reason is that with verified quotes we actually want to verify that the output the solver advertises is actually what the call data generates.
To do this we inject custom interactions into the execution plan that measures sell or buy token balances at critical points throughout the settlement. Of course we could inject these interactions in the driver but then the logic the injects the interactions and the logic that reads the balances is so far apart that it would just be very hard to follow.

I guess we could also make the driver return a fully formed settlement but then we'd have to inject the balance checking interactions into the encoded calldata. Which sounds doable but way more complicated and more effort than it's worth IMO.

File: crates/autopilot/src/arguments.rs
Comment (MartinquaXD) [reply crates/autopilot/src/arguments.rs:133]: I'm fine with splitting off these changes into individual PRs when the overall strategy of the solution is agreed upon and signed off on. Until then splitting out these PRs only to redo them 3 more times doesn't seem useful.
That's why I already made them part of this PR so we can see all the smaller changes this would entail in case there are more concerns.

File: crates/shared/src/price_estimation/trade_verifier.rs
Comment (MartinquaXD) [reply crates/shared/src/price_estimation/trade_verifier.rs:95]: Theoretically you can run into edge cases with pre-interactions.
For example you could request a quote for buying 2000 USDC for USDT using a pre-interaction that unstakes 2000 USDT from an AMM.
If we compute the `out_amount` as `usdt_before_settlement - usdt_after_settlement` our math won't make sense because it would look like we bought 2000 USDC for 0 USDT.
With our current logic we would check the USDT balance after running every pre-interation and before running any post-interactions (those don't matter for quoting anyway). This allows us to see that we actually spent the 2000 USDT we unstaked with the pre-interaction.

Comment (MartinquaXD) [reply crates/shared/src/price_estimation/trade_verifier.rs:95]: If post interactions are not an issue for quoting we could probably accomplish the same with a second simulation that only executes the pre interaction and measures the sell token balance before. And if the pre interaction somehow interferes with the proposed execution we'll still notice during the complete execution. Leaving just the trade preparations (e.g. approval, wrapping) to do in the API.

**Comment by MartinquaXD** [comment]
Synchronously discussed this and the conclusion was the we should first move what we have to prod (cautiously with sufficient logs and feature gates) and only improve upon that with internalization support once the original feature had impact on prod.

**Pull request closed without merging**
